{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import socket\n",
    "from socket import error as SocketError\n",
    "import errno\n",
    "import time\n",
    "from contextlib import suppress\n",
    "import traceback\n",
    "import struct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'connectionSocket' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-5-abd0408c0a0b>\u001b[0m in \u001b[0;36mmain\u001b[1;34m(wdir)\u001b[0m\n\u001b[0;32m     39\u001b[0m         \u001b[0mServerSocket\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0msocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msocket\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mAF_INET\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSOCK_STREAM\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 40\u001b[1;33m         \u001b[0mServerSocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbind\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"127.0.0.1\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m55556\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     41\u001b[0m         \u001b[0mServerSocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlisten\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mOSError\u001b[0m: [Errno 98] Address already in use",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-5-abd0408c0a0b>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m    108\u001b[0m     \u001b[0mpath\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdirname\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mspath\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    109\u001b[0m     \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchdir\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 110\u001b[1;33m     \u001b[0mmain\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-5-abd0408c0a0b>\u001b[0m in \u001b[0;36mmain\u001b[1;34m(wdir)\u001b[0m\n\u001b[0;32m     96\u001b[0m         \u001b[0mwlog\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Server closing on Error: \"\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0me\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m\" \"\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0me\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     97\u001b[0m         \u001b[0mwlog\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtb\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 98\u001b[1;33m         \u001b[0mconnectionSocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshutdown\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSHUT_RDWR\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     99\u001b[0m         \u001b[0mconnectionSocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    100\u001b[0m         \u001b[0mServerSocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshutdown\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSHUT_RDWR\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mUnboundLocalError\u001b[0m: local variable 'connectionSocket' referenced before assignment"
     ]
    }
   ],
   "source": [
    "class Logger():\n",
    "    def __init__(self,fullPath):\n",
    "        self.fpath=fullPath\n",
    "    def __call__(self,s):\n",
    "        with open(self.fpath,\"ab\") as fso:\n",
    "            s=time.strftime(\"%H:%M:%S -- %Y-%m-%d: \")+s\n",
    "            s+=\"\\n\"\n",
    "            s=s.encode()\n",
    "            fso.write(s)\n",
    "    def tb(self):\n",
    "        with open(self.fpath,\"a\") as fso:\n",
    "            (trash0,trash1,thetb)=sys.exc_info()\n",
    "            traceback.print_tb(thetb, file = fso)\n",
    "    def append(self,s):\n",
    "        with open(self.fpath,\"ab\") as fso:\n",
    "            s=s.encode()\n",
    "            fso.write(s)\n",
    "\n",
    "class RemoteSocketClosed(Exception):\n",
    "    pass\n",
    "    \n",
    "class OrderedToClose(Exception):\n",
    "    pass\n",
    "\n",
    "def justPrint(binary, connection):\n",
    "    print(binary.decode())\n",
    "\n",
    "def comeback(binary, connection):\n",
    "    binary= b\"'\"+binary+b\"' right back at ya!\"\n",
    "    print(binary)\n",
    "    connection.sendall(binary)\n",
    "\n",
    "def main(wdir):\n",
    "    processors=(justPrint,comeback)\n",
    "    key=b\"asdf\"\n",
    "    \n",
    "    wlog=Logger(os.path.join(wdir,\"log.txt\"))\n",
    "    try:\n",
    "        ServerSocket=socket.socket(socket.AF_INET, socket.SOCK_STREAM)\n",
    "        ServerSocket.bind((\"127.0.0.1\",55556))\n",
    "        ServerSocket.listen(0)\n",
    "        while True:\n",
    "            try:\n",
    "                (connectionSocket,address)=ServerSocket.accept()\n",
    "                connectionSocket.settimeout(30)\n",
    "                wlog.append(\"\\n\\n\")\n",
    "                wlog(\"connection to {} opened.\".format(address))\n",
    "                bInd=0\n",
    "                bLen=0\n",
    "                buff=b\"\"\n",
    "                mtype=None\n",
    "                mlen=None\n",
    "                headerRecieved=False\n",
    "                while True:\n",
    "                    if not headerRecieved:\n",
    "                        if bLen-bInd < 12:\n",
    "                            chunk=connectionSocket.recv(4096)\n",
    "                            if len(chunk)==0:\n",
    "                                raise RemoteSocketClosed\n",
    "                            bLen+=len(chunk)\n",
    "                            buff+=chunk\n",
    "                            continue\n",
    "                        if buff[bInd:bInd+4]==key:\n",
    "                            \n",
    "                            mtype, mlen = struct.unpack(\"<LL\",buff[bInd+4:bInd+12])\n",
    "                            bInd+=12\n",
    "                            headerRecieved=True\n",
    "                        else:\n",
    "                            print(1,\"error\")\n",
    "                            bInd+=1\n",
    "                    if headerRecieved:\n",
    "                        if bLen-bInd < mlen:\n",
    "                            chunk=connectionSocket.recv(4096)\n",
    "                            if len(chunk)==0:\n",
    "                                raise RemoteSocketClosed\n",
    "                            bLen+=len(chunk)\n",
    "                            buff+=chunk\n",
    "                            continue\n",
    "                        else:\n",
    "                            if mtype > len(processors)-1:\n",
    "                                raise ValueError(\"there is no message type \"+str(mtype))\n",
    "                            processors[mtype](buff[bInd:bInd+mlen],connectionSocket)\n",
    "                            buff=buff[bInd+mlen:]\n",
    "                            bInd=0\n",
    "                            bLen=len(buff)\n",
    "                            headerRecieved=False\n",
    "            except RemoteSocketClosed:\n",
    "                connectionSocket.shutdown(socket.SHUT_RDWR)\n",
    "                connectionSocket.close()\n",
    "                print(\"connection closed by remote socket\")\n",
    "            except socket.timeout:\n",
    "                connectionSocket.shutdown(socket.SHUT_RDWR)\n",
    "                connectionSocket.close()\n",
    "                print(\"connection timed out\")\n",
    "    except Exception as e:\n",
    "        wlog(\"Server closing on Error: \"+str(type(e))+\" \"+str(e))\n",
    "        wlog.tb()\n",
    "        connectionSocket.shutdown(socket.SHUT_RDWR)\n",
    "        connectionSocket.close()\n",
    "        ServerSocket.shutdown(socket.SHUT_RDWR)\n",
    "        ServerSocket.close()\n",
    "        print(9,buff)\n",
    "        raise e\n",
    "\n",
    "\n",
    "if __name__==\"__main__\":\n",
    "    spath=\"/home/mr/RPiProject/dummy.py\"\n",
    "    path=os.path.dirname(spath)\n",
    "    os.chdir(path)\n",
    "    main(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
